ferveo/src/refresh.rs

```diff
@@ -0,0 +1,384 @@
+            .iter()
+            .map(|p| {
+                // Current participant receives updates from other participants
+                let updates_for_participant: Vec<_> = share_updates
+                    .values()
+                    .map(|updates| *updates.get(p.index).unwrap())
+                    .collect();
+
+                // And updates their share
+                apply_updates_to_private_share::<E>(
+                    &p.private_key_share,
+                    &updates_for_participant,
+                )
+            })
+            .collect();
+
+        new_share_fragments
+    }
+
+    /// Ñ parties (where t <= Ñ <= N) jointly execute a "share recovery" algorithm, and the output is 1 new share.
+    /// The new share is intended to restore a previously existing share, e.g., due to loss or corruption.
+    #[test]
+    fn tdec_simple_variant_share_recovery_at_selected_point() {
+        let rng = &mut test_rng();
+        let shares_num = 16;
+        let threshold = shares_num * 2 / 3;
+
+        let (_, _, mut contexts) =
+            setup_simple::<E>(threshold, shares_num, rng);
+
+        // Prepare participants
+
+        // First, save the soon-to-be-removed participant
+        let selected_participant = contexts.pop().unwrap();
+        let x_r = selected_participant
+            .public_decryption_contexts
+            .last()
+            .unwrap()
+            .domain;
+        let original_private_key_share = selected_participant.private_key_share;
+
+        // Remove one participant from the contexts and all nested structures
+        let mut remaining_participants = contexts;
+        for p in &mut remaining_participants {
+            p.public_decryption_contexts.pop().unwrap();
+        }
+
+        // Each participant prepares an update for each other participant, and uses it to create a new share fragment
+        let new_share_fragments = make_new_share_fragments_for_recovery(
+            rng,
+            threshold,
+            &x_r,
+            &remaining_participants,
+        );
+
+        // Now, we have to combine new share fragments into a new share
+        let domain_points = &remaining_participants[0]
+            .public_decryption_contexts
+            .iter()
+            .map(|ctxt| ctxt.domain)
+            .collect::<Vec<_>>();
+        let new_private_key_share = recover_share_from_updated_private_shares(
+            &x_r,
+            &domain_points[..threshold],
+            &new_share_fragments[..threshold],
+        );
+
+        assert_eq!(new_private_key_share, original_private_key_share);
+
+        // If we don't have enough private share updates, the resulting private share will be incorrect
+        let incorrect_private_key_share =
+            recover_share_from_updated_private_shares(
+                &x_r,
+                &domain_points[..(threshold - 1)],
+                &new_share_fragments[..(threshold - 1)],
```

Comment:
We could throw in an `assert!` here to make sure the `domain_points` and `new_share_fragments` have a proper length. It's not strictly necessary, but it relates to some of the original issues we had with those tests.